src/solvers/nonlinear_solid_solver.cpp

```diff
@@ -0,0 +1,363 @@
+  int sc = height/2;
+  mfem::Vector v(vx.GetData() +  0, sc);
+  mfem::Vector x(vx.GetData() + sc, sc);
+  mfem::Vector dv_dt(dvx_dt.GetData() +  0, sc);
+  mfem::Vector dx_dt(dvx_dt.GetData() + sc, sc);
+
+  m_H_form->Mult(x, m_z);
+  m_S_form->TrueAddMult(v, m_z);
+  m_z.SetSubVector(m_ess_tdof_list, 0.0);
+  m_z.Neg(); // z = -z
+  m_M_solver.Mult(m_z, dv_dt);
+
+  dx_dt = v;
+}
+
+void NonlinearSolidDynamicOperator::ImplicitSolve(const double dt,
+    const mfem::Vector &vx, mfem::Vector &dvx_dt)
+{
+  int sc = height/2;
+  mfem::Vector v(vx.GetData() +  0, sc);
+  mfem::Vector x(vx.GetData() + sc, sc);
+  mfem::Vector dv_dt(dvx_dt.GetData() +  0, sc);
+  mfem::Vector dx_dt(dvx_dt.GetData() + sc, sc);
+
+  // By eliminating kx from the coupled system:
+  //    kv = -M^{-1}*[H(x + dt*kx) + S*(v + dt*kv)]
+  //    kx = v + dt*kv
+  // we reduce it to a nonlinear equation for kv, represented by the
+  // m_reduced_oper. This equation is solved with the m_newton_solver
+  // object (using m_J_solver and m_J_prec internally).
+  m_reduced_oper->SetParameters(dt, &v, &x);
+  mfem::Vector zero; // empty vector is interpreted as zero r.h.s. by NewtonSolver
+  m_newton_solver->Mult(zero, dv_dt);
+  MFEM_VERIFY(m_newton_solver->GetConverged(), "Newton solver did not converge.");
+  add(v, dt, dv_dt, dx_dt);
+}
+
+// destructor
+NonlinearSolidDynamicOperator::~NonlinearSolidDynamicOperator()
+{
+  delete m_M_mat;
+}
+
+
+NonlinearSolidReducedSystemOperator::NonlinearSolidReducedSystemOperator(
+  mfem::ParNonlinearForm *H_form, mfem::ParBilinearForm *S_form, mfem::ParBilinearForm *M_form,
+  const mfem::Array<int> &ess_tdof_list)
+  : mfem::Operator(M_form->ParFESpace()->TrueVSize()), m_M_form(M_form), m_S_form(S_form), m_H_form(H_form),
+    m_jacobian(nullptr), m_dt(0.0), m_v(nullptr), m_x(nullptr), m_w(height), m_z(height),
+    m_ess_tdof_list(ess_tdof_list)
+{ }
+
+void NonlinearSolidReducedSystemOperator::SetParameters(double dt, const mfem::Vector *v,
+    const mfem::Vector *x)
+{
+  m_dt = dt;
+  m_v = v;
+  m_x = x;
+}
+
+void NonlinearSolidReducedSystemOperator::Mult(const mfem::Vector &k, mfem::Vector &y) const
+{
+  // compute: y = H(x + dt*(v + dt*k)) + M*k + S*(v + dt*k)
+  add(*m_v, m_dt, k, m_w);
+  add(*m_x, m_dt, m_w, m_z);
+  m_H_form->Mult(m_z, y);
+  m_M_form->TrueAddMult(k, y);
+  m_S_form->TrueAddMult(m_w, y);
+  y.SetSubVector(m_ess_tdof_list, 0.0);
+}
+
+mfem::Operator &NonlinearSolidReducedSystemOperator::GetGradient(const mfem::Vector &k) const
+{
+  delete m_jacobian;
+  // Form the gradient of the complete nonlinear operator
+  mfem::SparseMatrix *localJ = Add(1.0, m_M_form->SpMat(), m_dt, m_S_form->SpMat());
+  add(*m_v, m_dt, k, m_w);
+  add(*m_x, m_dt, m_w, m_z);
+  localJ->Add(m_dt*m_dt, m_H_form->GetLocalGradient(m_z));
+  m_jacobian = m_M_form->ParallelAssemble(localJ);
+  delete localJ;
+
+  // Eliminate the fixed boundary DOFs
+  mfem::HypreParMatrix *Je = m_jacobian->EliminateRowsCols(m_ess_tdof_list);
```

Comment:
Can you add a comment that you don't need the internally malloc returned matrix?